docs/YieldDistributor.md

```diff
@@ -0,0 +1,245 @@
+
+3. It checks if the user has already voted in this cycle.
+
+4. If it's a new vote in the cycle, it resets the user's previous votes and adds their voting power to the current total votes.
+
+5. For each project:
+
+   - If it's a new vote, it initializes the user's distribution for that project.
+
+   - If it's an update to an existing vote, it subtracts the user's previous distribution.
+
+   - It calculates the new distribution based on the points allocated and the user's voting power.
+
+   - It updates both the overall project distribution and the user's personal distribution.
+
+6. It updates the last voted block number for the user.
+
+7. Finally, it emits an event with the voting details.
+
+
+Key points:
+- The function allows users to update their votes within a cycle.
+- It uses precision calculations to ensure accurate distribution of voting power.
+- It maintains both global project distributions and individual user distributions.
+- The voting power is based on the user's token holdings (both BREAD and BUTTERED_BREAD) over a specific period.
+
+## Voting power 
+```mermaid 
+sequenceDiagram
+    actor User
+    participant YD as YieldDistributor
+    participant BREAD as BREAD Token
+    participant BB as BUTTERED_BREAD Token
+
+    User->>YD: getCurrentVotingPower(account)
+    activate YD
+    YD->>YD: Calculate period start and end
+    YD->>YD: getVotingPowerForPeriod(BREAD, start, end, account)
+    activate YD
+    YD->>BREAD: numCheckpoints(account)
+    alt numCheckpoints == 0
+        YD-->>YD: Return 0
+    else numCheckpoints > 0
+        YD->>BREAD: checkpoints(account, 0)
+        alt first checkpoint > end
+            YD-->>YD: Return 0
+        else first checkpoint <= end
+            loop Find latest checkpoint within interval
+                YD->>BREAD: checkpoints(account, index)
+            end
+            YD->>YD: Calculate initial voting power
+            loop Process remaining checkpoints
+                YD->>BREAD: checkpoints(account, index)
+                YD->>YD: Update total voting power
+                alt checkpoint <= start
+                    YD->>YD: Adjust for interval start
+                    YD-->>YD: Break loop
+                end
+            end
+        end
+    end
+    YD-->>YD: Return BREAD voting power
+    deactivate YD
+    
+    YD->>YD: getVotingPowerForPeriod(BUTTERED_BREAD, start, end, account)
+    activate YD
+    Note over YD: Same process as BREAD
+    YD-->>YD: Return BUTTERED_BREAD voting power
+    deactivate YD
+    
+    YD->>YD: Sum BREAD and BUTTERED_BREAD voting power
+    YD-->>User: Return total voting power
+    deactivate YD
+```
+Now, let's break down the getVotingPowerForPeriod function step by step:
+1. Input Validation:
+    - Check if the start time is before the end time.
+    - Ensure the end time is not after the current block.
+2. Initial Checkpoint Check:
+    - Get the total number of checkpoints for the account.
```

Comment:
what is a checkpoint? why would an account have checkpoints? i think this is dealing with the part that scales voting power according to how long the account has been holding the BREAD, but that wouldnt be clear to me from reading at this point